### PR TITLE
Add Hover provider

### LINKED ISSUE #350: Add Hover Provider
Opening issue just to say I'm working on adding Hover to lexicon. I'm working on my fork here: https://github.com/bkanuka/lexicon

I'm working on the tests/recordings and then I'll open a PR

### PR DESCRIPTION
**Pull request by bkanuka**
Closes #350

### CODE REVIEW

**Comment by CapsuleCD** [comment]
Hi.

I'm an automated pull request bot named [CapsuleCD](https://www.capsulecd.com). I handle testing, versioning and package releases for this project. 
- If you're the owner of this repo, you can click the button below to kick off the CapsuleCD build pipeline and create an automated release.'
- If not, don't worry, someone will be by shortly to check on your pull request. 

[CapsuleCD](https://www.capsulecd.com/project/github/AnalogJ/lexicon/pullrequests/351)

---
If you're interested in learning more about [CapsuleCD](http://www.github.com/AnalogJ/capsulecd), or adding it to your project, you can check it out [here](https://www.capsulecd.com)

**Comment by bkanuka** [comment]
I think the CI is failing on things I didn't touch. Let me know if there's anything I should change!

**Comment by adferrand** [comment]
Yes, a deprecated option has been finally removed from pip 19.0. Hopefully I anticipated the problem with #338. The PR has been merged. You can sync your PR with master to fix the CI for you.

**Comment by bkanuka** [comment]
Okay. I rebased and tests are still failing but this time it's my fault :smile:

**Comment by adferrand** [comment]
Hello @bkanuka! Yes vcrpy can be tricky to configure. So some basics first. I think you understood that vcrpy is mocking the http requests, in order to avoid real http calls. It will return, for each matching request in the cassette, the http response that have been registered.

There is one cassette per test. Each cassette must contain each request and response that will be done during the test, and in the correct order.

A given request will match a registered request if the url (including params), the body, and the http headers, corresponds in what is emitted and what is registered. Likewise, all data retrieved for the registered response will be the only data available to the logic that is been tested.

**Comment by adferrand** [comment]
So basically, the development workflow cycle consists in these steps:
* remove all cassettes for your providers in `tests/fixtures/cassettes/hover`
* execute `LEXICON_LIVE_TESTS=true py.test tests/providers/test_hover.py` to trigger vcrpy in record mode: actual http calls will be done, and all exchanges will be recorded in the cassettes
* execute ``LEXICON_LIVE_TESTS=false py.test tests/providers/test_hover.py` to trigger vcrpy in replay mode: this time, no actual call is made, everything is get from the cassettes

At this time, if all tests are passing, the cassettes are technically ready to be committed. Indeed, upon commit, the CI will run directly vcrpy in replay mode, using the cassettes, and passing the test.

But obviously, if no filter is done, every sensible information is stored in the cassettes. So you need to filter them out. 

However, keep in mind this workflow. It is what you need to do for every modification you will do on the filters from now on to validate the cassettes before commit.

**Comment by adferrand** [comment]
Now I described all this context, here is the troubleshooting process I propose to you.

First, execute the validation cycle I described before, without any filter. It will check that there is no problem with the principle of vcrpy cassettes itself for your case. 

Then, reexecute this validation cycle by adding, one by one, the params to filter in the relevant filter functions.

When you have a particular param filtered that breaks the cassettes, switch this param to the tupple declaration I described before, to replace its value to a fixed dummy value.

If this dummy value breaks some logic, adapt the dummy value to be accepted by the logic (eg. a particular string length, or a pattern to respect).

Hopefully it will be sufficient, and you will end with cassettes without sensible materials, that work in replay mode. 

Then you can commit all of this. 

If not, post here the specific param that is not handled correctly, and we will figure out alternate ways to mock it.

**Comment by bkanuka** [comment]
@adferrand nevermind! I was running the recording/tests from the wrong directory on my machine. Silly mistake.  Still having issues but I think I can figure it out. Then I'll do a rebase, clean up my git history, and ping you again when things are ready to merge.

**Comment by adferrand** [comment]
Yes, I was on the point to say it to you ^^

Also, merge latest master upstream to ensure to have the latest optimizations I made on lexicon test system.

**Comment by bkanuka** [comment]
@adferrand Okay everything should be good now! Thanks for your help!

**Comment by adferrand** [comment]
Great! I start the review.

File: lexicon/tests/providers/test_hover.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                filter_body['domains'] = [
+                    {
+                        'id': self.domain_id,
+                        'domain_name': self.domain,
+                        'status': 'active',
+                    },
+                ]
+
+            response['body']['string'] = json.dumps(filter_body).encode('UTF-8')
+        return response
```
Comment (adferrand) [inline lexicon/tests/providers/test_hover.py]: Can you add a newline here ?

File: lexicon/providers/hover.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        if query_params is None:
+            query_params = {}
+        response = requests.request(action, self.api_endpoint + url,
+                                    params=query_params,
+                                    data=json.dumps(data),
+                                    cookies=self.cookies,
+                                    headers={'Content-Type': 'application/json'})
+        # if the request fails for any reason, throw an error.
+        response.raise_for_status()
+        return response.json()
```
[the remark this answers is not in the document]
Comment (adferrand) [reply lexicon/providers/hover.py]: Also, can you add a newline here?

**Comment by bkanuka** [comment]
In reply to all your comments about json:
> Are we sure that Hover API will always return a JSON ?

I mean, yes, until the API changes - and then we'll have bigger problems anyway. I can throw an Exception if the response is not JSON but there won't be any graceful way of handling that.

File: lexicon/tests/providers/test_hover.py (hunk 1 above)
Comment (bkanuka) [reply lexicon/tests/providers/test_hover.py]: wait, lint failed when I added a trailing newline. Do you want a trailing/blank newline or no?

Comment (adferrand) [reply lexicon/tests/providers/test_hover.py]: Sorry I was not on the last line... I was talking of the classic blank newline you put at the end to make some editors happy, but it was already there. So you can ignore the comment on do not add anything.

**Comment by bkanuka** [comment]
Okay review/accept as you please!

**Comment by adferrand** [comment]
LGTM! Thanks a lot @bkanuka!

**Pull request merged**